File: NSTouchBarCatalogCreatingandCustomizingNSTouchBar/Objective-C/NSTouchBar Catalog/WindowController.m
/*
 Copyright (C) 2017 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information
 
 Abstract:
 Main window controller for this sample.
 */

#import "WindowController.h"

static NSTouchBarItemIdentifier WindowControllerLabelIdentifier = @"com.TouchBarCatalog.windowController.label";
static NSString *backgroundWindowIdentifier = @"BackgroundWindow";

@interface WindowController () <NSTouchBarDelegate>

// Background Window used to test the TouchBar in the context of an NSPopover.
@property (nonatomic, strong) NSWindowController *backgroundWindowController;

@end


#pragma mark -

@implementation WindowController

- (void)windowDidLoad
{
    [super windowDidLoad];
    
    self.window.frameAutosaveName = @"WindowAutosave";

    // Load the Background Window from it's separate storyboard.
    NSStoryboard *storyboard = [NSStoryboard storyboardWithName:backgroundWindowIdentifier bundle: nil];
    _backgroundWindowController = [storyboard instantiateControllerWithIdentifier:backgroundWindowIdentifier];
    self.backgroundWindowController.window.frameAutosaveName = backgroundWindowIdentifier;
    [self.backgroundWindowController showWindow:nil];
}

#pragma mark NSTouchBarProvider

// This window controller will have only one NSTouchBarItem instance, which is a simple label,
// so to show that view controller bar can reside along side its window controller.
//
- (NSTouchBar *)makeTouchBar
{
    NSTouchBar *bar = [[NSTouchBar alloc] init];
    bar.delegate = self;
        
    // Set the default ordering of items.
    bar.defaultItemIdentifiers =
        @[WindowControllerLabelIdentifier, NSTouchBarItemIdentifierOtherItemsProxy];
    
    return bar;
}

#pragma mark NSTouchBarDelegate

// This gets called while the NSTouchBar is being constructed, for each NSTouchBarItem to be created.
- (nullable NSTouchBarItem *)touchBar:(NSTouchBar *)touchBar makeItemForIdentifier:(NSTouchBarItemIdentifier)identifier
{
    if ([identifier isEqualToString:WindowControllerLabelIdentifier])
    {
        NSTextField *theLabel = [NSTextField labelWithString:NSLocalizedString(@"Catalog", @"")];
        
        NSCustomTouchBarItem *customItemForLabel =
            [[NSCustomTouchBarItem alloc] initWithIdentifier:WindowControllerLabelIdentifier];
        customItemForLabel.view = theLabel;
        
        // We want this label to always be visible no matter how many items are in the NSTouchBar instance.
        customItemForLabel.visibilityPriority = NSTouchBarItemPriorityHigh;
        
        return customItemForLabel;
    }
    
    return nil;
}

@end

File: NSTouchBarCatalogCreatingandCustomizingNSTouchBar/Swift/NSTouchBar Catalog/WindowController.swift
/*
 Copyright (C) 2017 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information
 
 Abstract:
 Main window controller for this sample.
 */

import Cocoa

struct AssetNames {
    static let image = "Sunset_2"
    static let red = "Red"
    static let green = "Green"
}

// MARK: -

@available(OSX 10.12.2, *)
fileprivate extension NSTouchBarItem.Identifier {
    static let label = NSTouchBarItem.Identifier("com.TouchBarCatalog.TouchBarItem.label")
}

// MARK: -

class WindowController: NSWindowController {
    
    let backgroundWindowIdentifier = "BackgroundWindow"
    var backgroundWindowController: NSWindowController!
    
    override func windowDidLoad() {
        super.windowDidLoad()
        
        window?.setFrameAutosaveName(NSWindow.FrameAutosaveName(rawValue: "WindowAutosave"))
        
        // Load the Background Window from it's separate storyboard.
        let storyboard = NSStoryboard(name:NSStoryboard.Name(rawValue: backgroundWindowIdentifier), bundle: nil)
        if let windowController =
            storyboard.instantiateController(withIdentifier: NSStoryboard.SceneIdentifier(rawValue: backgroundWindowIdentifier))
                as? NSWindowController {
            backgroundWindowController = windowController
            backgroundWindowController.window?.setFrameAutosaveName(NSWindow.FrameAutosaveName(rawValue: backgroundWindowIdentifier))
            backgroundWindowController.showWindow(nil)
        }
    }
    
    // MARK: NSTouchBarProvider
    
    @available(OSX 10.12.2, *)
    override func makeTouchBar() -> NSTouchBar? {
        let touchBar = NSTouchBar()
        touchBar.delegate = self
        touchBar.customizationIdentifier = NSTouchBar.CustomizationIdentifier("com.TouchBarCatalog.windowTouchBar")
        touchBar.defaultItemIdentifiers = [.label, .fixedSpaceLarge, .otherItemsProxy]
        
        return touchBar
    }
    
}

// MARK: - NSTouchBarDelegate

@available(OSX 10.12.2, *)
extension WindowController: NSTouchBarDelegate {
    
    // This gets called while the NSTouchBar is being constructed, for each NSTouchBarItem to be created.
    func touchBar(_ touchBar: NSTouchBar, makeItemForIdentifier identifier: NSTouchBarItem.Identifier) -> NSTouchBarItem? {
        switch identifier {
        case NSTouchBarItem.Identifier.label:
            let custom = NSCustomTouchBarItem(identifier: identifier)
            let label = NSTextField(labelWithString: NSLocalizedString("Catalog", comment: ""))
            custom.view = label
            
            return custom
            
        default:
            return nil
        }
    }
}

